feat(@schematics/angular): cli app redesign

sjtrimble opened the pull request:
- Updated ng new template design
- Update favicon to be black as default

CC: @IgorMinar

sjtrimble commented:
### Initial Mockups

**Desktop**
![ngnew-app-desktop-rocket](https://user-images.githubusercontent.com/2958442/57564582-2880d880-7363-11e9-9af9-92979668abd1.png)

**Mobile**
![ngnew-app-mobile](https://user-images.githubusercontent.com/2958442/57564593-45b5a700-7363-11e9-8092-64a2f91c2850.png)

filipesilva commented:
@alexeagle do you know if this should be part of CLI version 8?

filipesilva commented on packages/schematics/angular/application/files/src/index.html.template:
```diff
@@ -7,6 +7,311 @@
 
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
+  <link
```
I feel users might just forget these here when removing the demo code.

I tried moving them into `src/styles.css` like this:
```
@import "https://fonts.googleapis.com/css?family=Roboto";
@import "https://fonts.googleapis.com/css?family=Roboto+Mono";
@import "https://fonts.googleapis.com/icon?family=Material+Icons";
```

And it does work, in Chrome at least. But it's not converted to anything else. It relies on native browser support for `@import` in CSS. This it wouldn't be a great choice either. Might just be better to keep it here.

sjtrimble replied:
I see your point about it not being obvious for users to delete. Like with the rest of the CSS I like having it all in one place to just focus on one file when removing template code.

What if we add comments for now that say "remove this template code"? I'll implement that for now until we come up with something better.

sjtrimble replied:
Added the comments `<!-- *** Remove Template Code Below *** -->` and `<!-- *** Template Code Ends *** -->` in both template and in index files.

filipesilva replied:
That SGTM, let's roll with it.

sjtrimble commented on packages/schematics/angular/application/files/src/index.html.template:
```diff
@@ -7,6 +7,311 @@
+    />
+
+  <style>
+    body {
```
The idea was to keep them in the `index` file (even though not really the way things are normally done) so that the user can easily delete all of the template code in one full swoop.

sjtrimble commented on packages/schematics/angular/application/other-files/app.component.html.template:
```diff
@@ -1,21 +1,231 @@
+
+    </a>
+
+    <a class="circle-link" title="Find a Local Meetup" href="https://www.meetup.com/topics/angularjs/" target="_blank" rel="noopener">
```
We should probably use the keyword link 👍 Like you, I was also not able to find "just Angular" as a topic group.

sjtrimble replied:
Updated the link to https://www.meetup.com/find/?keywords=angular.

alexeagle commented:
Looks great to get this into 8, but needs to land by tomorrow

sjtrimble commented on packages/schematics/angular/application/other-files/app.component.html.template:
```diff
@@ -1,21 +1,231 @@
+    <div class="card card-small" (click)="selection.value = 'test'" tabindex="0">
+      <svg class="material-icons" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"/></svg>
+
+      <span>Run Test</span>
```
Updated text to say 'Run and Watch Tests'

sjtrimble commented:
@alexeagle let me know if there is anything you need from me! I just made the updates and rebased.

filipesilva commented:
@sjtrimble I'll pick it up from here and fix the failing tests. They are mostly related to the unit tests and e2e expecting the old structure.

filipesilva commented:
Followup PR in https://github.com/angular/angular-cli/pull/14484

alexeagle commented:
@sjtrimble could we remove the changes to index.html? We think developers will accidentally leave these behind in their application.

1) do we really need the custom font? can we avoid using material icons and fall back to browser built-in font?
2) can we inline as many styles as possible?
3) if styles are needed on the body, could they be injected by the component constructor

we want it to be sufficient to delete the component and then you have a clean starting point for an app without any of this boilerplate left behind

alan-agius4 commented:
For fonts maybe we can use `@import` in the components style file? 
See support: https://developer.mozilla.org/en-US/docs/Web/CSS/@import

IgorMinar commented on packages/schematics/angular/application/files/src/index.html.template:
```diff
@@ -7,6 +7,315 @@
+      -moz-osx-font-smoothing: grayscale;
+    }
+
+    body,
```
could we move all of these stylesheets into the component itself? We can't rely on people to correctly remove the style sheet from their app.

You can use https://next.angular.io/guide/component-styles#host-context to get the sheets into the top level document stylesheet.

And for the fonts or anything else that can't be added via that method, we could just inject the https://next.angular.io/api/common/DOCUMENT into the component constructor and do manual dom manipulation from there to create and append fonts to the document.head.

sjtrimble commented:
Copied over #14484 test fixes per @IgorMinar

sjtrimble commented:
Moved all code to the template file for easy removal, except for body zero padding/margin styles.

IgorMinar commented:
Let's get this done in 8.1. I'm sorry for deferring it but there are still questions to resolve and as much as I love this change, we shouldn't rush it and pollute new projects with code that we'll regret in the future.

vthinkxie commented:
any update on this, it seems that 8.1 has not included this feat

mgechev commented:
Let's see if we can release it as part of version 8.2.

vikerman commented:
Hi @sjtrimble - We discussed how to take this forward.

We would like to include this as part of the 8.2 release.

We would like the starter to have the following properties:
- No changes to index.html
- All changes the user need to do while changing it to their are restricted to  `app.component.html`, `app.component.css`
- Sort of simple CSS for the app component. The current one is very long
- Use a system font without needing to link to an external font file

sjtrimble commented:
1. ✅I rebased on the latest master so everything is fresh.
2. ✅I removed the external font references/links and replaced with system fonts.
3. :small_orange_diamond: Let's chat about the CSS - I think there is value in having a presentable and visually appealing first impression but perhaps we can simplify the code a bit.
4. :small_orange_diamond: I had managed to take everything out of the `index.html` file already *EXCEPT* the body styles - I don't know how else we would access those? Maybe we can use negative margins? I have to check this out again.

sjtrimble commented:
@vikerman FYI ⬆️

IgorMinar commented on packages/schematics/angular/application/files/src/index.html.template:
```diff
@@ -7,6 +7,7 @@
 
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
+
```
remove?

sjtrimble replied:
I assume you mean the extra line space here?

IgorMinar commented on packages/schematics/angular/application/other-files/app.component.html.template:
```diff
@@ -1,21 +1,506 @@
+        <path id="Path_19" data-name="Path 19" d="M4137.529,1027.283l-7.7-3.723,4.417-2.721,7.753,3.723Z" transform="translate(-4125.003 -1015.199)" fill="#fff"/>
+      </svg>
+
+      <!-- <img src="../assets/animations.png" height="20px" alt="Angular Animations Logo"/> -->
```
commented out?

IgorMinar commented on packages/schematics/angular/application/other-files/app.component.html.template:
```diff
@@ -1,21 +1,506 @@
+
+</div>
+
+<!-- *** Template Code Ends *** -->
```
as mentioned above, this should be a highly visible multiline comment/banner as well so that it's easy to identify. Also at the top you used "placeholder" terminology and down here you are using "Template", can you please make it consistent? thanks!

vikerman commented on packages/schematics/angular/application/other-files/app.component.html.template:
```diff
@@ -1,21 +1,505 @@
+<!-- * * * * * * * * * * is only a placeholder * * * * * * * * * * -->
+<!-- * * * * * * * * * * and can be replaced. * * * * * * * * * * * -->
+<!-- * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * -->
+<!-- * * * * * * * * * Delete the template below * * * * * * * * * * -->
```
It's not below :) We can probably leave this sentence out in this part.

vikerman replied:
Or maybe just say - End of placeholder

sjtrimble replied:
Ahhh, I said above on the first line and forgot the last sentence :) I like your idea of end of placeholder better! Thanks :)
